### DISCUSSION TITLE
is DataPropertyCanOnlyHaveOneType neccesary?

### DISCUSSION THREAD

**Discussion by quyle92** [spatie/laravel-data#229] [General]
Hi,
Method `getValueForProperty()` of `EmptyDataResolver` class is thrown Exception in the scenario where property has more than 1 type:
```
if ($property->types()->count() > 1) {
            throw DataPropertyCanOnlyHaveOneType::create($property);
}
```
I wonder why we don't just remove that Exception and let make the first type as default when calling `empty()`? as I have multiple properties with 2 types so whenever calling `empty()`, I have to manually specify the default type for each which is somewhat inconvenient.

**Comment by rubenvanassche** [comment]
We actually base ourselves a lot on this rule, that a property can only have one type (besides the Optional, Lazy and null types). For example in the case of empty, a type like this would provide breaking values:

```
int|array // should return null or []
```

We also depend on the one type rule to build up our validation rules, so again:

```
int|array // should return ['number'] or ['array']
```

That's why we're not going to allow more than one type in the future.

**Reply by quyle92** [reply]
ok thanks.... but I am a bit confused... isn't php 8's union type allowed more than 1 type? if so, how does it cause breaking value?

**Reply by rubenvanassche** [reply]
You're indeed allowed to use multiple types for one property within PHP 8. But we need to make some assumptions in this package to let it work correctly and that's why we've decided to add a rule you can only define one type for a property (besides the package internal Optional and Lazy type).
